scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -877,10 +877,16 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     def compoundTypeRest(typ: Type): Type = {
       val startPos = typ.startTokenPos
       var t = typ
-      while (acceptOpt[KwWith]) {
+      // Indentation means a refinement and we cannot join refinements this way
+      while (acceptOpt[KwWith] && !token.is[Indentation.Indent]) {
         val rhs = annotType()
         t = atPos(startPos, rhs)(Type.With(t, rhs))
       }
+      token match {
+        case Indentation.Indent() =>
```

Comment:
can indentation come **without** a `with`? the code allows that.

Reply:
We shouldn't have an indetation without a `with`, but it's a good idea to make sure!

Reply:
Added a test and refactored the function to use tail recursion. This way we don't check any tokens twice.